### PR TITLE
Introduce HTTP client for GeoserverInterceptorService

### LINKED ISSUE #25: Introduce HTTP client util for GeoServerInterceptor
A client util should be provided in order to generate interceptor rules easily via HTTP.

Follow up to #22

### PR DESCRIPTION
**Pull request by ahennr**
This PR introduces a base Java client for SHOGun REST interfaces (AutoClosable which can be used in `try-resources`-blocks) with factory class `ShogunClientBuilder`. Currently, only GeoServer-interceptor rules can be modified.

For example, a manager client for GeoServer interceptor can be created in the following way:
```java
String username = "interceptoradmin";
String password = "<<YOUR_ADMIN_PASSWORD>>";
String baseUrl = "https://url-to-shogun/shogun-gs-interceptor";

try (
    ShogunGsInterceptorManager manager = (ShogunGsInterceptorManager) ShogunClientBuilder.builder()
        .password(password)
        .username(username)
        .shogunServiceBaseUrl(baseUrl)
        .managerType(ShogunManagerType.GEOSERVER_INTERCEPTOR)
        .build()
    ) {

    // fetch all interceptor rules
    List<InterceptorRule> rules = manager.getAllInterceptorRules();

    // add request rule
    String endpoint = "TOP";
    boolean success = manager.addRequestRuleForServiceAndEndpoint(endpoint, OgcEnum.ServiceType.WFS, RuleType.DENY);
}  catch (Exception e) {
  System.err.println(e);
}
```
Closes #25 

Please review @terrestris/devs

### CODE REVIEW

File: shogun-gs-interceptor/src/main/java/de/terrestris/shogun/interceptor/security/access/entity/InterceptorRulePermissionEvaluator.java [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+@Component
+public class InterceptorRulePermissionEvaluator extends BaseEntityPermissionEvaluator<InterceptorRule> {
+
+    @Autowired
+    protected SecurityContextUtil securityContextUtil;
+
+    @Override
+    public boolean hasPermission(User user, InterceptorRule entity, PermissionType permission) {
+        List<GrantedAuthority> authorities = securityContextUtil.getGrantedAuthorities();
+        boolean isInterceptorAdmin = authorities.stream().anyMatch(
```
Comment (LukasLohoff) [inline shogun-gs-interceptor/src/main/java/de/terrestris/shogun/interceptor/security/access/entity/InterceptorRulePermissionEvaluator.java:27]: maybe we can write a util for this, which can check roles for a given authorities collection
basically hasRole() or hasAnyRole() as a directly invokable method

or [isUserInRole](https://docs.spring.io/spring-security/site/apidocs/org/springframework/security/web/servletapi/SecurityContextHolderAwareRequestWrapper.html#isUserInRole(java.lang.String)) can be used, but not sure if that works

File: shogun-manager/src/main/java/de/terrestris/shogun/manager/AbstractShogunManager.java [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+     * Closes the underlying client.
+     */
+    @Override
+    public void close() {
+        closed = true;
+        try {
+            httpClient.close();
+        } catch (IOException e) {
+            // TODO
+            e.printStackTrace();
```
Comment (LukasLohoff) [inline shogun-manager/src/main/java/de/terrestris/shogun/manager/AbstractShogunManager.java:197]: Missing exception handling

**Review by LukasLohoff — APPROVED** [review]
LGTM! I left some general minor suggestions (logging, utils). I don't have much experience with java clients, so additional reviews are welcome

File: shogun-manager/src/main/java/de/terrestris/shogun/manager/AbstractShogunManager.java [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+     */
+    private byte[] handleOkResult(CloseableHttpResponse response) {
+        try (
+            ByteArrayOutputStream byteArrayOutputStream = new ByteArrayOutputStream()
+        ) {
+            HttpEntity iStream = response.getEntity();
+            iStream.writeTo(byteArrayOutputStream);
+            return byteArrayOutputStream.toByteArray();
+        } catch (Exception e) {
+            log.error("Could not parse result to resulting byte array");
```
Comment (dnlkoch) [inline shogun-manager/src/main/java/de/terrestris/shogun/manager/AbstractShogunManager.java:181]:
```suggestion
            log.error("Could not parse result to resulting byte array: {}", e.getMessage());
```

File: shogun-manager/src/main/java/de/terrestris/shogun/manager/AbstractShogunManager.java [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+            log.trace(e);
```
Comment (dnlkoch) [inline shogun-manager/src/main/java/de/terrestris/shogun/manager/AbstractShogunManager.java:182]:
```suggestion
            log.trace("Full stack trace: {}", e);
```

File: shogun-manager/src/main/java/de/terrestris/shogun/manager/ShogunGsInterceptorManager.java [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+        }
+    }
+
+    /**
+     * Remove all interceptor rules for given endpoint
+     * @param endpoint The endpoint
+     * @return true if successful, false otherwise
+     * @throws Exception The exception
+     */
+    public boolean removeAllRulesForEndpoint(String endpoint) throws Exception {
```
Comment (dnlkoch) [inline shogun-manager/src/main/java/de/terrestris/shogun/manager/ShogunGsInterceptorManager.java:69]: Absolutely minor, but I prefer that delete methods should not have a return value.

File: shogun-manager/src/main/java/de/terrestris/shogun/manager/ShogunGsInterceptorManager.java [hunk 6]
```diff
@@ -0,0 +1,205 @@
...
+
+    /**
+     * Add interceptor <i>request</i> rule for given service and endpoint
+     * @param endpoint The endpoint to add the rule for
+     * @param service The {@link de.terrestris.shogun.interceptor.enumeration.OgcEnum.ServiceType}
+     * @param rule The {@link RuleType}
+     * @return true if successful, false otherwise
+     * @throws Exception The exception
+     */
+    public boolean addRequestRuleForServiceAndEndpoint(String endpoint, OgcEnum.ServiceType service, RuleType rule) throws Exception {
```
Comment (dnlkoch) [inline shogun-manager/src/main/java/de/terrestris/shogun/manager/ShogunGsInterceptorManager.java:88]: For create methods I would prefer to return the created entity.

File: shogun-manager/src/main/resources/log4j2.yml [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+        Pattern: '%highlight{%d{HH:mm:ss.SSS} %-5level %logger{36}.%M() @%L - %msg%n}{FATAL=red blink, ERROR=red, WARN=yellow bold, INFO=white, DEBUG=green bold, TRACE=blue}'
+  Loggers:
+    Root:
+      level: info
+      AppenderRef:
+        ref: STDOUT
+
+    logger:
+      - name: de.terrestris
+        level: trace
```
Comment (dnlkoch) [inline shogun-manager/src/main/resources/log4j2.yml:15]:
```suggestion
        level: info
```

**Review by dnlkoch — APPROVED** [review]
Very nice! Thanks!

File: shogun-gs-interceptor/src/main/java/de/terrestris/shogun/interceptor/security/access/entity/InterceptorRulePermissionEvaluator.java (hunk 1 above)
Comment (ahennr) [reply shogun-gs-interceptor/src/main/java/de/terrestris/shogun/interceptor/security/access/entity/InterceptorRulePermissionEvaluator.java:27]: Good idea. I'll introduce this in an upcoming PR

**Pull request merged**